### PR TITLE
Support for reviewers & assignees

### LINKED ISSUE #215: Add options to specify labels, reviewers and assignees to be associated with the created pull request
It would be very helpful to be able to specify labels, reviewers and assignees for the pull request that gets created by the gradle plugin.

### PR DESCRIPTION
**Pull request by StefMa**
fixes #215 

This time the forgotten reviewers and assignees 🙂 

Haven't tested it yet ...

### CODE REVIEW

File: src/main/java/org/gradle/wrapperupgrade/UpgradeWrapper.java [hunk 1]
```diff
@@ -201,6 +203,36 @@ private void gitCreatePr(Params params, String prTitle, String prBody) throws IO
...
+            List<String> reviewers = upgrade.getOptions().getReviewers().get();
+            if (!reviewers.isEmpty()) {
+                List<GHUser> githubReviewers = reviewers.stream()
+                    .map(reviewer -> {
+                        try {
+                            return params.gitHub.getUser(reviewer);
+                        } catch (IOException e) {
+                            getLogger().warn(String.format("Error fetching GitHub user '%s'", reviewer), e);
+                            return null;
+                        }
```
Comment (alextu) [inline src/main/java/org/gradle/wrapperupgrade/UpgradeWrapper.java:215]: Suggesting to extract as a function and re-use for the assignees

**Review by alextu — APPROVED** [review]
Thanks for your contribution! LGTM, just suggesting a small refactoring and we can merge it once you confirm it works properly (I'll test it too)

File: src/main/java/org/gradle/wrapperupgrade/UpgradeWrapper.java [hunk 2]
```diff
@@ -201,6 +203,14 @@ private void gitCreatePr(Params params, String prTitle, String prBody) throws IO
             if (!labels.isEmpty()) {
                 pr.addLabels(labels.toArray(new String[0]));
             }
+            List<String> reviewers = upgrade.getOptions().getReviewers().get();
+            if (!reviewers.isEmpty()) {
+                List<GHUser> githubReviewers = mapToGHUsers(params, reviewers);
```
Comment (alextu) [inline src/main/java/org/gradle/wrapperupgrade/UpgradeWrapper.java:208]: I think you forgot to apply the reviewers to the PR with `pr.requestReviewers(githubReviewers)`, same for the assignees

File: src/main/java/org/gradle/wrapperupgrade/UpgradeWrapper.java (hunk 1 above)
Comment (StefMa) [reply src/main/java/org/gradle/wrapperupgrade/UpgradeWrapper.java]: Done here https://github.com/gradle/wrapper-upgrade-gradle-plugin/pull/271/commits/d4cbfb935661209acf1961d03a742b1348e10798

**Comment by StefMa** [comment]
> once you confirm it works properly 

I can confirm that "it works". 

Assignees works perfectly. 
See https://github.com/StefMa/wrapper-upgrade-gradle-plugin/pull/6

I “couldn’t test” reviewers. First, I tried to put you as a reviewer, then I got the message from GitHub (via the GitHub API project) that you can only add collaborators as reviewers. Then I set myself as a reviewer, but that didn’t work either. The message was that you can’t set the reviewer to the same person as the author.

Because those errors indicate that there are API calls happening, I would assume that everything works as expected.

However, those errors throw an exception and therefore failed the build. Maybe we want to catch those and report them only via the logger? 🤔 What do you think?

**Comment by alextu** [comment]
> > once you confirm it works properly
> 
> I can confirm that "it works".
> 
> Assignees works perfectly. See [StefMa#6](https://github.com/StefMa/wrapper-upgrade-gradle-plugin/pull/6)
> 
> I “couldn’t test” reviewers. First, I tried to put you as a reviewer, then I got the message from GitHub (via the GitHub API project) that you can only add collaborators as reviewers. Then I set myself as a reviewer, but that didn’t work either. The message was that you can’t set the reviewer to the same person as the author.
> 
> Because those errors indicate that there are API calls happening, I would assume that everything works as expected.
> 
> However, those errors throw an exception and therefore failed the build. Maybe we want to catch those and report them only via the logger? 🤔 What do you think?

I've tested it as well and noticed the same, I'll be able to test it end to end tomorrow. I agree that we should not fail the build since the PR is created, and just output a warning message in that case.

**Comment by StefMa** [comment]
I changed the code a bit and cleaned it up.
Check it out here: https://github.com/gradle/wrapper-upgrade-gradle-plugin/pull/271/commits/a2a408eb2350564d90a330e525aecc8f552b9b80

* Add try-catch around `addAssignees` and `requestReviewers` and `addLabels`
* In case of an error, we just log it as warning
* Also I moved the parts after the output of "PR created", because actually this happens afterwards
* I extract methods for each of them 

I tested the case of "request review for myself (author)" and the build succeeded, but the error is visible:
Screenshot 2024-11-29 at 8 29 03 AM

I tested to add a label that **doesn't exist** to check if an error happen.
But actually it just creates that label 🫠 
See https://github.com/StefMa/wrapper-upgrade-gradle-plugin/pull/8 🤷 
Anyways, according to the [method signature](https://github.com/hub4j/github-api/blob/107c8e6efe5c9d350bf6632302bce633bf548291/src/main/java/org/kohsuke/github/GHIssue.java#L408) it could also throw an `IOException`.
So better safe than sorry I would say.

**Comment by alextu** [comment]
Thanks! I've tested it with reviewers as well, looks good, I'll merge the PR soon

**Review by alextu — APPROVED** [review]

**Pull request merged**